rust/template/differential_datalog/program.rs

```diff
@@ -1662,8 +1716,12 @@ impl Program {
                     while cursor.val_valid(&storage) && *cursor.key(&storage) == k {
                         let mut weight = 0;
                         cursor.map_times(&storage, |_, diff| weight += diff);
-                        assert!(weight >= 0);
-                        if weight > 0 {
+                        //assert!(weight >= 0);
```

Comment:
I think that what you are saying here is that we don't have a semantics for operations that combine sets and multisets. This is the most important thing to figure out. Ideally we would be very explicit about this, giving users control by using either distinct or a conversion in the other way, that casts a set to a multiset preserving weights. I was a little uncomfortable about implementing sets with multisets even previously, because we never had a proof that we achieve the expected datalog semantics by inserting distinct in some places. Is this something that the Vienna guys would be interested in working on?

Reply:
Actually, I don't see how this combination is problematic.  Inputs are always multisets, it's just that sometimes we force them to have unit multiplicities (before feeding data to differential).

This particular code is just me being unwilling to change the semantics of indexes from sets to multisets.